flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(rawTimer.f0, outputView);
+			LongSerializer.INSTANCE.serialize(rawTimer.f1, outputView);
+			keySerializer.serialize(rawTimer.f2, outputView);
+			namespaceSerializer.serialize(rawTimer.f3, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while serializing the raw timer.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private Tuple4<Integer, Long, K, N> deserializeRawTimer(byte[] bytes) {
+		ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
+		DataInputViewStreamWrapper inputView = new DataInputViewStreamWrapper(inputStream);
+
+		try {
+			int keyGroup = IntSerializer.INSTANCE.deserialize(inputView);
+			long timestamp = LongSerializer.INSTANCE.deserialize(inputView);
+			K key = keySerializer.deserialize(inputView);
+			N namespace = namespaceSerializer.deserialize(inputView);
+
+			return new Tuple4<>(keyGroup, timestamp, key, namespace);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the raw timer.", e);
+		}
+	}
+
+	/**
+	 * A timer store backed by RocksDB. 
+	 * 
+	 * The timers are stored in RocksDB in the order of key groups. To allow 
+	 * efficient access, the timers are partitioned and the leader of each 
+	 * partition is stored in an in-memory heap. The top of the heap is 
+	 * exactly the first timer to trigger. The heap is updated whenever the 
+	 * partition's leader is updated.
+	 */
+	private class PersistentTimerHeap {
+
+		private final ColumnFamilyHandle handle;
+
+		/** Leader timers in the partitions */
+		private final Tuple4<Integer, Long, K, N>[] partitionHeadTimers;
+		
+		/** The order of the partition's leader in the heap */
+		private final int[] orders;
+		
+		/** The partitions in the order of their leaders in the heap */
+		private final int[] indices;
+		
+		@SuppressWarnings("unchecked")
+		PersistentTimerHeap(int capacity, ColumnFamilyHandle handle) {
+			this.handle = handle;
+
+			this.partitionHeadTimers = (Tuple4<Integer, Long, K, N>[])new Tuple4[capacity];
+			this.indices = new int[capacity];
+			this.orders = new int[capacity];
+			
+			for (int i = 0; i < capacity; i++) {
+				orders[i] = i;
+				indices[i] = i;
+			}
+		}
+
+		private int getParentOrder(int order) {
+			return (order - 1) / 2;
+		}
+
+		private int getLeftChildOrder(int order) {
+			return order * 2 + 1;
+		}
+
+		private int getRightChildOrder(int order) {
+			return order * 2 + 2;
+		}
+
+		private long getTimestamp(Tuple4<Integer, Long, K, N> rawTimer) {
+			return rawTimer == null ? Long.MAX_VALUE : rawTimer.f1;
+		}
+
+		/** Rebuild the heap with the given DB. */
+		void initialize() {
+			int currentPartition = -1;
+			Tuple4<Integer, Long, K, N> currentPartitionLeaderTimer = null;
+
+			RocksIterator iterator = db.newIterator(handle);
```

Comment:
I suggest using try-with-resource with the `RocksIterator` to ensure that the native resources are closed after usage. This can also be changed in other places.